### PR TITLE
Add Timeout property to Web component

### PR DESCRIPTION
**Pull request by austenjoa**
Allows for users to handle Web request timeouts through two methods; a TimeOut block for each Web component, as well as through the Screen error handler. Addresses feature request #1782.

@ewpatton

### CODE REVIEW

**Comment by AppInventorWorkerBee** [comment]
Can one of the admins verify this patch?

**Comment by ewpatton** [comment]
@austenjoa A couple things that also need to be changed:

- [ ] Update the version number of the Web component defined in YaVersion.java and leave a comment there describing what you've added. See the previous updates for examples.
- [ ] Update YoungAndroidFormUpgrader.java to handle the new version number (no changes need to be made since you're just adding a property and event)
- [ ] Update versioning.js to handle the new version number (specify 'noUpgrade' for the aforementioned reason).
- [ ] Describe the new property and event in the component documentation at docs/reference/components/connectivity.html

**Comment by austenjoa** [comment]
@ewpatton Updated the documentation accordingly.

**Comment by ewpatton** [comment]
@AppInventorWorkerBee ok to test

File: appinventor/docs/reference/components/connectivity.html [hunk 1]
```diff
@@ -707,6 +707,8 @@ <h3>Properties</h3>
   <dd>Whether the response should be saved in a file.</dd>
   <dt><code>Url</code></dt>
   <dd>The URL for the web request.</dd>
+  <dt><code>Timeout</code></dt>
+  <dd>The number of milliseconds that the Web component will wait for a response to a web request. If 0, then the Web component will wait infinitely.</dd>
```
Comment (ewpatton) [inline appinventor/docs/reference/components/connectivity.html]: infinitely => indefinitely?

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java [hunk 2]
```diff
@@ -929,6 +1006,7 @@ private static HttpURLConnection openConnection(CapturedProperties webProps, Str
       throws IOException, ClassCastException, ProtocolException {
 
     HttpURLConnection connection = (HttpURLConnection) webProps.url.openConnection();
+    connection.setReadTimeout(webProps.timeout);
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: I think we want to set both `setConnectionTimeout` and `setReadTimeout`.

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java [hunk 3]
```diff
@@ -887,8 +956,7 @@ private void performRequest(final CapturedProperties webProps, byte[] postData,
         processResponseCookies(connection);
 
         if (saveResponse) {
-          final String path = saveResponseContent(connection, webProps.responseFileName,
-              responseType);
+          final String path = saveResponseContent(connection, webProps.responseFileName, responseType);
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: I think IntelliJ reformatted this line on you. Revert this change.

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java [hunk 4]
```diff
@@ -122,6 +124,21 @@
     }
   }
 
+  /**
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: It's possible we might want to reuse this exception for other classes that do network operations. Let's move it into `com.google.appinventor.components.runtime.errors`

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java [hunk 5]
```diff
@@ -909,6 +977,15 @@ public void run() {
           });
         }
 
+      } catch (SocketTimeoutException e){
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Add a space between `)` and `{`

File: appinventor/docs/reference/components/connectivity.html (hunk 1 above)
Comment (austenjoa) [reply appinventor/docs/reference/components/connectivity.html]: Changed accordingly for next commit

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java (hunk 3 above)
Comment (austenjoa) [reply appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Fixed this

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java (hunk 5 above)
Comment (austenjoa) [reply appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Updated

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java (hunk 4 above)
Comment (austenjoa) [reply appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Moved it into it's own class.

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java (hunk 2 above)
Comment (austenjoa) [reply appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Set both props

**Comment by austenjoa** [comment]
@ewpatton Updated accordingly

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java [hunk 6]
```diff
@@ -325,6 +331,31 @@ public void ResponseFileName(String responseFileName) {
     this.responseFileName = responseFileName;
   }
 
+  /**
+   * Returns the number of milliseconds that each request will wait for a response before they time out.
+   * If set to 0, then the request will wait for a response indefinitely.
+   */
+  @SimpleProperty(category = PropertyCategory.BEHAVIOR,
+      description = "The number of milliseconds that a web request will wait for a response before giving up." +
+              "If set to 0, then there is no time limit on how long the request will wait.")
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Indentation here should be +4 from the `d` in `description`.

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java [hunk 7]
```diff
@@ -364,6 +395,9 @@ public void run() {
         } catch (FileUtil.FileException e) {
           form.dispatchErrorOccurredEvent(Web.this, METHOD,
               e.getErrorMessageNumber());
+        } catch (RequestTimeoutException e) {
+          form.dispatchErrorOccurredEvent(Web.this, METHOD,
+                  ErrorMessages.ERROR_WEB_REQUEST_TIMED_OUT, webProps.urlString);
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Indentation here should be +4 from `f` in `form` (under `.`).  Also make this update for the other places in this file.

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java (hunk 6 above)
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Add a space before `If` or after `.` on the previous line. Otherwise, they end up concatenated together in the block tooltip.

Comment (austenjoa) [reply appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Fixed

Comment (austenjoa) [reply appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Fixed this

File: appinventor/components/src/com/google/appinventor/components/runtime/Web.java (hunk 7 above)
Comment (austenjoa) [reply appinventor/components/src/com/google/appinventor/components/runtime/Web.java]: Corrected all of these

**Comment by austenjoa** [comment]
@ewpatton Made the requested whitespace changes
